src/Swarm/TUI/Inventory/Sorting.hs

```diff
@@ -0,0 +1,79 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- Module      :  Swarm.TUI.Inventory.Sorting
+-- Copyright   :  Brent Yorgey
+-- Maintainer  :  byorgey@gmail.com
+--
+-- SPDX-License-Identifier: BSD-3-Clause
+module Swarm.TUI.Inventory.Sorting (
+  InventorySortOptions (..),
+  InventorySortDirection (..),
+  InventorySortOrder (..),
+  cycleSortOrder,
+  cycleSortDirection,
+  sortInventory,
+  renderSortMethod,
+) where
+
+import Control.Lens (view)
+import Data.List (sortBy)
+import Data.Ord (comparing)
+import Data.Text qualified as T
+
+import Algorithms.NaturalSort (sortKey)
+
+import Swarm.Game.Entity as E
+
+data InventorySortDirection
+  = Ascending
+  | Descending
+  deriving (Enum, Bounded, Eq)
+
+data InventorySortOrder
+  = ByNaturalAlphabetic
+  | ByQuantity
+  | ByType
+  deriving (Enum, Bounded, Eq)
+
+data InventorySortOptions = InventorySortOptions InventorySortDirection InventorySortOrder
+
+renderSortMethod :: InventorySortOptions -> T.Text
+renderSortMethod (InventorySortOptions direction order) =
+  T.unwords [prefix, label]
+ where
+  prefix = case direction of
+    Ascending -> "↑"
+    Descending -> "↓"
+  label = case order of
+    ByNaturalAlphabetic -> "name"
+    ByQuantity -> "count"
+    ByType -> "type"
+
+succCyclic :: (Enum a, Bounded a, Eq a) => a -> a
+succCyclic x =
+  if x == maxBound
+    then minBound
+    else succ x
+
+cycleSortOrder :: InventorySortOptions -> InventorySortOptions
+cycleSortOrder (InventorySortOptions direction order) =
+  InventorySortOptions direction (succCyclic order)
+
+cycleSortDirection :: InventorySortOptions -> InventorySortOptions
+cycleSortDirection (InventorySortOptions direction order) =
+  InventorySortOptions (succCyclic direction) order
+
+getSortCompartor :: Ord a => InventorySortOrder -> (a, Entity) -> (a, Entity) -> Ordering
+getSortCompartor order = case order of
+  ByNaturalAlphabetic -> comparing $ sortKey . T.toLower . view entityName . snd
+  ByQuantity -> comparing fst
+  ByType -> comparing $ view entityProperties . snd
```

Comment:
The same comment goes for sorting by type: among entities with the same type the order seems very random, I think it would be nice to secondarily sort by name.